search/locality_scorer.hpp

```diff
@@ -37,29 +43,36 @@ class LocalityScorer
 private:
   struct ExLocality
   {
-    ExLocality();
-    explicit ExLocality(Locality const & locality);
+    ExLocality() = default;
+    ExLocality(Locality const & locality, double queryNorm, uint8_t rank);
 
-    inline uint32_t GetId() const { return m_locality.m_featureId; }
+    uint32_t GetId() const { return m_locality.m_featureId; }
 
     Locality m_locality;
-    size_t m_numTokens;
-    uint8_t m_rank;
-    NameScore m_nameScore;
+    double m_queryNorm = 0.0;
+    double m_similarity = 0.0;
+    uint8_t m_rank = 0;
   };
 
   friend std::string DebugPrint(ExLocality const & locality);
 
   // Leaves at most |limit| elements of |localities|, ordered by some
   // combination of ranks and number of matched tokens.
-  void LeaveTopLocalities(size_t limit, std::vector<Locality> & localities) const;
+  void LeaveTopLocalities(IdfMap & idfs, size_t limit, std::vector<Locality> & localities) const;
+
+  // Selects at most |limitUniqueIds| best features by query norm and
+  // rank, and then leaves only localities corresponding to those
+  // features in |els|.
+  void LeaveTopByNormAndRank(size_t limitUniqueIds, std::vector<ExLocality> & els) const;
 
-  void RemoveDuplicates(std::vector<ExLocality> & ls) const;
-  void LeaveTopByRankAndProb(size_t limit, std::vector<ExLocality> & ls) const;
-  void SortByNameAndProb(std::vector<ExLocality> & ls) const;
+  // Leaves at most |limit| best localities by similarity to the query
```

Comment:
As a variant:
Leaves at most |limit| unique best localities by similarity to the query and rank.

Reply:
Done